compiler_gym/views/observation_space_spec.py

```diff
@@ -77,6 +77,12 @@ def __init__(
         self.translate = translate
         self.to_string = to_string
 
+    def __hash__(self) -> int:
+        # Quickly hash observation spaces by comparing the index into the list
+        # of spaces returned by the environment. This means that hashing across
+        # different environments is _not_ safe.
+        return self.index
```

Comment:
> is _not_ safe.
Why not safe? I can see it as not being ideal because it might cause more collisions than expected, but I'm not sure why it's not safe.

Also, why not hash with the id (and the index if needed). If it has to be faster than string hashing, then cache it?

Reply:
Good points. I'll clarify that comment and hash a (str, int) tuple